subprojects/core/src/main/java/org/gradle/api/attributes/Attribute.java

```diff
@@ -117,4 +119,30 @@ public int hashCode() {
     public String toString() {
         return name;
     }
+
+    private static <T> void validateType(String name, Class<T> clazz) {
+        if (clazz.isArray()) {
+            validateType(name, clazz.getComponentType());
+            return;
+        }
+        if (Scalars.isScalarType(clazz)) {
+            return;
+        }
+        if (Named.class.isAssignableFrom(clazz)) {
```

Comment:
❌ I can implement the `Named` interface using any kind of mutable class. We'll need to be strciter here and check if it is one of our auto-generated named types which are known to be immutable.